File: backend/projects.py
"""Projects: named workspaces that own annotations, classes, and a trained head (Phase 7).

A PROJECT is one experiment. Two projects over the same slides train independently
and predict differently, so "glands vs stroma" and "tumor vs normal" can coexist
instead of contaminating a single pooled classifier (which is what happened before
this module existed: one global head, trained on every slide on disk).

WHAT A PROJECT OWNS — and, just as importantly, what it does NOT:

    data/projects/<project_id>/
        project.json                 # name, slides, classes
        annotations/<slide_id>.json  # the regions drawn IN THIS PROJECT
        models/head__<embed_id>.*    # the head trained on THIS PROJECT's annotations

    NOT owned (deliberately global, shared by every project):
        data/slides/           the WSI files themselves
        data/cache/features/   THE FEATURE BANK
        data/cache/            DeepZoom tiles

The feature bank is the one that matters. It is keyed by (slide, embedding model,
tissue mask) — nothing about a user's experiment enters it — so it is a derived
property of a FILE, not of a project. Sharing it means a slide swept once with
Virchow 2 (minutes of GPU) is reused by every project that contains it. Scoping it
per project would re-run the single most expensive step in the whole pipeline, once
per project, for zero gain.

TWO INVARIANTS. Both are load-bearing; violating either reintroduces a bug class we
specifically designed away.

  1. `project.json["slides"]` is a DISPLAY LIST. It decides what the slide picker
     shows. It NEVER decides what the model trains on — training iterates this
     project's annotations/ directory (see `annotated_slides`). Without this rule you
     get "I removed the slide from the project but the head still knows about it".

  2. Annotation tags are the SOURCE OF TRUTH for training; `project.json["classes"]`
     is only a vocabulary + colour registry. So deleting a class is purely cosmetic:
     its annotations keep their tag and keep training, and the head reports them as
     `classes_not_in_project` so the UI can offer to re-add the class.

     Corollary: there is deliberately NO class rename. A label is a free-text string
     inside every annotation that uses it, so renaming would mean rewriting every
     annotation file in the project — a migration inside a migration. Add / remove /
     recolour only.

The project id is a slug of the name ("Glands vs stroma" -> "glands-vs-stroma") so
that `ls data/projects/` is readable. THE ID IS FROZEN AT CREATION: renaming changes
only the display name, never the id or the directory.

This module imports ONLY `slides` (for REPO_ROOT and the traversal-guard pattern). It
must never import classifier / jobs / features / annotations — that is what keeps it
out of the existing features -> jobs -> classifier -> features import cycle.
"""

from __future__ import annotations

import json
import os
import re
import shutil
import threading
from datetime import datetime
from pathlib import Path

from backend import slides

PROJECTS_DIR = slides.REPO_ROOT / "data" / "projects"

# Serialises "delete a project" against "train a project". Without it, a train that
# is already in flight can re-create the directory tree it is writing into (joblib
# needs the dir to exist), and a project you just deleted partially comes back from
# the dead. jobs.py holds this across a training pass; delete() takes it too.
# Re-entrant because delete() calls other functions in this module that also take it.
LOCK = threading.RLock()

# The class vocabulary a NEW project starts with. Lifted from the frontend, which
# used to own this list (annotate.js PRESET_CLASSES + CLASS_PALETTE). It lives here
# now because a class carries an explicit stored COLOUR, and colour has to survive a
# browser cache clear.
DEFAULT_CLASSES = [
    "gland",
    "epithelium",
    "stroma",
    "tumor",
    "necrosis",
    "lymphocytes",
    "blood vessel",
    "adipose",
    "background",
]

# Distinct and reasonably colourblind-friendly. A new class takes the first colour
# NOT already in use (see `_next_color`) rather than the one at its index — indexing
# by position meant deleting a class silently repainted every class after it.
DEFAULT_PALETTE = [
    "#e6194B", "#3cb44b", "#4363d8", "#f58231", "#911eb4",
    "#42d4f4", "#f032e6", "#bfef45", "#fabed4", "#469990",
    "#9A6324", "#800000", "#808000", "#000075", "#a9a9a9",
]

_COLOR_RE = re.compile(r"^#[0-9a-fA-F]{6}$")


# --- Ids and paths -----------------------------------------------------------

def _slug(name: str) -> str:
    """Turn a display name into a filesystem- and URL-safe id."""
    s = re.sub(r"[^a-z0-9]+", "-", str(name).strip().lower()).strip("-")
    return s or "project"


def _safe_dir(project_id: str) -> Path | None:
    """The directory a project id MAPS TO, whether or not it exists yet.

    THE TRAVERSAL GUARD. `delete()` runs shutil.rmtree on the path this returns, and
    the id arrives from a URL path segment — so this function is the only thing
    standing between a malformed request and an arbitrary directory on disk. It
    mirrors slides._resolve_slide_path: reject anything that isn't a plain name, then
    prove the resolved path sits DIRECTLY inside PROJECTS_DIR.
    """
    if not project_id or "/" in project_id or "\\" in project_id or project_id.startswith("."):
        return None
    path = PROJECTS_DIR / project_id
    # `resolve()` collapses any "..", so this catches traversal even if the checks
    # above were somehow bypassed. A symlinked project dir also fails here.
    if path.resolve().parent != PROJECTS_DIR.resolve():
        return None
    return path


def project_dir(project_id: str) -> Path | None:
    """The directory of an EXISTING project, or None (bad id, or no such project)."""
    path = _safe_dir(project_id)
    if path is None or not (path / "project.json").exists():
        return None
    return path


def annotations_dir(project_id: str) -> Path | None:
    path = project_dir(project_id)
    return None if path is None else path / "annotations"


def models_dir(project_id: str) -> Path | None:
    path = project_dir(project_id)
    return None if path is None else path / "models"


# --- Read --------------------------------------------------------------------

def load(project_id: str) -> dict | None:
    """Parse a project's project.json, or None if there is no such project.

    Read straight from disk every time. These files are ~1 KB and the app is local;
    an mtime-checked cache would be a bug factory for no measurable gain.
    """
    path = project_dir(project_id)
    if path is None:
        return None
    try:
        with (path / "project.json").open("r", encoding="utf-8") as fh:
            return json.load(fh)
    except (json.JSONDecodeError, OSError):
        return None


def list_projects() -> list[dict]:
    """Every project on disk, oldest first. Summary shape, for the picker."""
    if not PROJECTS_DIR.exists():
        return []
    out = []
    for path in sorted(PROJECTS_DIR.iterdir()):
        if not path.is_dir():
            continue
        project = load(path.name)
        if project is None:
            continue
        out.append({
            "id": project["id"],
            "name": project["name"],
            "created_at": project.get("created_at", ""),
            "n_slides": len(project.get("slides", [])),
            "n_classes": len(project.get("classes", [])),
        })
    out.sort(key=lambda p: p["created_at"])
    return out


def projects_with_slide(slide_id: str) -> list[str]:
    """Ids of the projects containing `slide_id`. Used by the end-of-sweep retrain hook:
    fresh vectors for a slide invalidate the head of EVERY project that uses it."""
    return [p["id"] for p in list_projects()
            if slide_id in (load(p["id"]) or {}).get("slides", [])]


def annotated_slides(project_id: str) -> list[str]:
    """The slides this project has an annotation file for — i.e. WHAT TRAINING READS.

    Deliberately NOT project["slides"] (invariant 1): membership is a display list.
    The training set is defined by the annotation files that actually exist, so adding
    or removing a slide from the picker can never silently change what the head learnt.
    """
    path = annotations_dir(project_id)
    if path is None or not path.exists():
        return []
    return sorted(p.stem for p in path.iterdir() if p.suffix == ".json")


def has_annotations(project_id: str, slide_id: str) -> bool:
    """True if this project holds a NON-EMPTY annotation file for the slide."""
    path = annotations_dir(project_id)
    if path is None:
        return False
    file = path / f"{slide_id}.json"
    if not file.exists():
        return False
    try:
        with file.open("r", encoding="utf-8") as fh:
            return len(json.load(fh)) > 0
    except (json.JSONDecodeError, OSError):
        return False


# --- Write -------------------------------------------------------------------

def save(project: dict) -> dict:
    """Overwrite a project's project.json atomically.

    Same tmp + os.replace dance as annotations.py: an interrupted write must never
    leave a half-written project.json, because that would read back as "no such
    project" and hide the user's annotations.
    """
    path = _safe_dir(project["id"])
    if path is None:
        raise ValueError(f"unsafe project id {project['id']!r}")
    path.mkdir(parents=True, exist_ok=True)

    target = path / "project.json"
    tmp = target.with_suffix(".json.tmp")
    with tmp.open("w", encoding="utf-8") as fh:
        json.dump(project, fh, indent=2)
    os.replace(tmp, target)
    return project


def _next_color(classes: list[dict]) -> str:
    """The first palette colour not already taken, so adding a class never re-tints
    an existing one. Wraps once the palette is exhausted."""
    used = {c.get("color", "").lower() for c in classes}
    for color in DEFAULT_PALETTE:
        if color.lower() not in used:
            return color
    return DEFAULT_PALETTE[len(classes) % len(DEFAULT_PALETTE)]


def _clean_classes(raw: list) -> list[dict]:
    """Validate + normalise a class list. Raises ValueError on bad input.

    Names are unique case-insensitively (they're matched against annotation tags,
    which are free text); colours must be #rrggbb. A class with no colour gets the
    next unused palette entry.
    """
    out: list[dict] = []
    seen: set[str] = set()
    for entry in raw:
        if isinstance(entry, str):
            entry = {"name": entry}
        if not isinstance(entry, dict):
            raise ValueError(f"class must be a name or {{name, color}}, got {entry!r}")

        name = str(entry.get("name", "")).strip()
        if not name:
            raise ValueError("class name cannot be empty")
        if name.lower() in seen:
            raise ValueError(f"duplicate class name {name!r}")
        seen.add(name.lower())

        color = str(entry.get("color", "")).strip()
        if not color:
            color = _next_color(out)
        elif not _COLOR_RE.match(color):
            raise ValueError(f"colour must be #rrggbb, got {color!r}")

        out.append({"name": name, "color": color})
    return out


def create(name: str, slide_ids: list[str] | None = None,
           classes: list | None = None) -> dict:
    """Create a project. Ids collide-suffix (-2, -3) and are frozen thereafter.

    `slide_ids` defaults to every slide on disk — the common case is one experiment
    over everything you have, and it means a fresh project never starts with orphaned
    slides that are invisible in the picker.
    """
    with LOCK:
        base = _slug(name)
        project_id, n = base, 1
        while (PROJECTS_DIR / project_id).exists():
            n += 1
            project_id = f"{base}-{n}"

        if slide_ids is None:
            slide_ids = [s["id"] for s in slides.list_slides()]

        project = {
            "id": project_id,
            "name": str(name).strip() or project_id,
            "created_at": datetime.now().isoformat(timespec="seconds"),
            "slides": list(slide_ids),
            "classes": _clean_classes(classes if classes is not None else DEFAULT_CLASSES),
        }
        save(project)
        # Create both subdirectories up front so every other module can assume they
        # exist and never has to mkdir into a project (which is how a deleted project
        # comes back from the dead — see LOCK).
        (PROJECTS_DIR / project_id / "annotations").mkdir(parents=True, exist_ok=True)
        (PROJECTS_DIR / project_id / "models").mkdir(parents=True, exist_ok=True)
        return project


def rename(project_id: str, name: str) -> dict | None:
    """Change the DISPLAY NAME only. The id and directory never move — every stored
    URL, every localStorage key, and every path on disk stays valid."""
    with LOCK:
        project = load(project_id)
        if project is None:
            return None
        new_name = str(name).strip()
        if not new_name:
            raise ValueError("project name cannot be empty")
        project["name"] = new_name
        return save(project)


def delete(project_id: str) -> dict | None:
    """Remove a project and EVERYTHING IT OWNS — and nothing else.

    Gone:  data/projects/<pid>/  (project.json, its annotations, its heads)
    Kept:  data/slides/**        the WSIs
           data/cache/**         the shared feature bank + DeepZoom tiles

    Returns a receipt of what was removed, so a caller (or a test) can assert on it.
    """
    with LOCK:
        path = project_dir(project_id)
        if path is None:
            return None

        n_annotations = len(annotated_slides(project_id))
        models = models_dir(project_id)
        n_heads = len(list(models.glob("head__*.joblib"))) if models.exists() else 0

        # Belt and braces before an rmtree driven by a URL path segment. project_dir()
        # already proved this, but the cost of being wrong here is somebody's home
        # directory, so we prove it again immediately before the destructive call.
        resolved = path.resolve()
        if resolved.parent != PROJECTS_DIR.resolve():
            raise ValueError(f"refusing to delete {resolved} — outside {PROJECTS_DIR}")
        shutil.rmtree(resolved)

        return {"deleted": project_id,
                "removed": {"annotations": n_annotations, "heads": n_heads}}


def set_classes(project_id: str, classes: list) -> dict | None:
    """Replace the whole class list (same 'send the whole collection' convention as
    annotations — one convention, learned once). Raises ValueError on bad input."""
    with LOCK:
        project = load(project_id)
        if project is None:
            return None
        project["classes"] = _clean_classes(classes)
        return save(project)


def add_slide(project_id: str, slide_id: str) -> dict | None:
    """Put a slide in this project's picker. No-op if it's already there."""
    with LOCK:
        project = load(project_id)
        if project is None:
            return None
        if slide_id not in project["slides"]:
            project["slides"].append(slide_id)
            save(project)
        return project


def migrate_legacy() -> dict:
    """One-time import of the pre-Phase-7 layout into a project called "default".

    Before projects existed, annotations lived in a flat data/annotations/ and the one
    global head in data/models/. This lifts them into data/projects/default/ so an
    existing install keeps its work and behaves exactly as it did (the default project
    contains every slide on disk, which is what the old global head trained on).

    COPY, NEVER MOVE — and guarded by a SENTINEL FILE, not by "does data/projects/
    exist?". That distinction is the whole design:

        The naive trigger resurrects data. Create projects, delete them all, restart:
        data/projects/ is empty again, so the migration re-runs and the annotations you
        deliberately deleted come back from the dead.

    The sentinel is written LAST, so a half-finished migration simply runs again. And
    because we copy, data/annotations/ and data/models/ survive untouched as a frozen
    backup — which also makes `rm -rf data/projects/` a clean, complete "start over".

    Idempotent. Safe to call on every boot. Returns a small report for the log.
    """
    with LOCK:
        sentinel = PROJECTS_DIR / ".legacy_imported"
        if sentinel.exists():
            return {"status": "already_imported"}

        legacy_annotations = slides.REPO_ROOT / "data" / "annotations"
        legacy_models = slides.REPO_ROOT / "data" / "models"

        PROJECTS_DIR.mkdir(parents=True, exist_ok=True)
        project = load("default") or create("Default")

        copied_annotations = copied_heads = 0
        if legacy_annotations.exists():
            target = PROJECTS_DIR / project["id"] / "annotations"
            target.mkdir(parents=True, exist_ok=True)
            for src in sorted(legacy_annotations.glob("*.json")):
                dst = target / src.name
                if dst.exists():
                    continue          # never overwrite work already in the project
                shutil.copy2(src, dst)
                copied_annotations += 1

        if legacy_models.exists():
            target = PROJECTS_DIR / project["id"] / "models"
            target.mkdir(parents=True, exist_ok=True)
            for src in sorted(legacy_models.glob("head__*")):
                dst = target / src.name
                if dst.exists():
                    continue
                shutil.copy2(src, dst)
                copied_heads += 1

        sentinel.write_text(
            f"legacy data/annotations + data/models imported into "
            f"'{project['id']}' at {datetime.now().isoformat(timespec='seconds')}\n"
            "Delete data/projects/ entirely to re-run this import.\n",
            encoding="utf-8",
        )
        return {"status": "imported", "project_id": project["id"],
                "annotations": copied_annotations, "heads": copied_heads}


def remove_slide(project_id: str, slide_id: str, force: bool = False) -> dict | None:
    """Take a slide out of this project's picker.

    Returns {"status": "has_annotations", ...} instead of acting when the project
    still holds annotations for it and `force` is false — a mis-click must not be
    able to silently destroy drawn work. With force=True we delete THIS PROJECT'S
    annotation file for the slide and nothing else: never the slide, never the bank,
    never another project's annotations.
    """
    with LOCK:
        project = load(project_id)
        if project is None:
            return None

        if has_annotations(project_id, slide_id) and not force:
            return {"status": "has_annotations", "slide_id": slide_id}

        if force:
            annos = annotations_dir(project_id)
            if annos is not None:
                (annos / f"{slide_id}.json").unlink(missing_ok=True)

        project["slides"] = [s for s in project["slides"] if s != slide_id]
        save(project)
        return {"status": "ok", "project": project}

File: frontend/js/projects.js
// Projects: named workspaces, and the app's BOOT DRIVER (Phase 7).
//
// A project owns its annotations, its class list, and its own trained head, so two
// experiments over the same slides can coexist instead of contaminating one pooled
// classifier. It does NOT own the slides (files on disk) or the feature bank — the
// bank is keyed by (slide, embedding model, tissue mask) and is deliberately SHARED,
// so a slide swept once with Virchow 2 is reused by every project containing it.
//
// THE EVENT CONTRACT (three lines; the rest of the frontend depends on them):
//
//   slideprobe:project-opened   "forget everything about the old project."
//                               Handlers must be SYNCHRONOUS STATE RESETS ONLY — they
//                               must not open a slide or start long work.
//                               detail = { project: {id, name, slides, classes} }
//
//   slideprobe:slide-opened     "this slide is now on screen; load its stuff."
//                               Unchanged, EXCEPT slideId is now `null` when the
//                               active project has no slides. Every consumer must
//                               treat null as "nothing on screen".
//
//   slideprobe:classes-changed  the class list/colours changed WITHIN this project.
//                               detail = { classes }. Deliberately not a re-dispatch
//                               of project-opened, which would needlessly re-open the
//                               slide and flicker the viewer.
//
// WHY THIS FILE OWNS BOOT, and why script order alone could not fix it:
// viewer.js used to auto-open slides[0] on DOMContentLoaded. Registering our listener
// first would NOT have helped — ours is async, so it yields at its first `await` and
// viewer's handler runs immediately after, on the same tick. So viewer.js no longer
// boots itself: it exposes SlideViewer.showProjectSlides() and we CALL it, after
// dispatchEvent has returned. dispatchEvent is synchronous, so by then every module
// has finished resetting. Two strict phases, no timers, no setTimeout(…, 0).

(function () {
  const ACTIVE_KEY = "slideprobe.activeProject";

  let ALL_SLIDES = [];   // every slide on disk (global; the "add slides" source)
  let PROJECTS = [];     // the summary list, for the picker
  let ACTIVE = null;     // the full active project: {id, name, slides, classes}

  const el = (id) => document.getElementById(id);

  function setProjStatus(msg, isError) {
    const s = el("projects-status");
    if (!s) return;
    s.textContent = msg || "";
    s.className = isError ? "err" : "";
  }

  // --- Colour -----------------------------------------------------------------

  // A label that isn't (yet) a project class still has to be drawn in SOMETHING. Hash
  // its NAME to a colour, never its position in a list: position is exactly what the
  // old colorForClass() used, which is why deleting one class silently repainted every
  // class after it.
  const FALLBACK_PALETTE = [
    "#e6194B", "#3cb44b", "#4363d8", "#f58231", "#911eb4",
    "#42d4f4", "#f032e6", "#bfef45", "#fabed4", "#469990",
  ];

  function colorOf(name) {
    const key = String(name || "").trim().toLowerCase();
    const hit = (ACTIVE ? ACTIVE.classes : []).find(
      (c) => c.name.toLowerCase() === key
    );
    if (hit) return hit.color;
    let h = 0;
    for (const ch of key) h = (h * 31 + ch.charCodeAt(0)) >>> 0;
    return FALLBACK_PALETTE[h % FALLBACK_PALETTE.length];
  }

  const classes = () => (ACTIVE ? ACTIVE.classes.slice() : []);
  const classNames = () => classes().map((c) => c.name);

  // --- Class edits (the project is the source of truth; the server assigns colours) --

  async function putClasses(next) {
    if (!ACTIVE) return;
    try {
      ACTIVE.classes = await API.setClasses(ACTIVE.id, next);
      renderPanel();
      document.dispatchEvent(
        new CustomEvent("slideprobe:classes-changed", {
          detail: { classes: classes() },
        })
      );
    } catch (err) {
      setProjStatus(err.message, true);
    }
  }

  // Add a class if it's new (case-insensitive). A blank `color` tells the server to
  // pick the first UNUSED palette entry.
  async function addClass(name) {
    const trimmed = String(name || "").trim();
    if (!trimmed || !ACTIVE) return false;
    if (classNames().some((c) => c.toLowerCase() === trimmed.toLowerCase())) return false;
    await putClasses([...classes(), { name: trimmed, color: "" }]);
    return true;
  }

  // Removing a class is purely COSMETIC. Annotation tags are the source of truth for
  // training, so regions labelled with it keep their tag and keep training — they just
  // fall back to a hashed colour until the class is added back.
  async function removeClass(name) {
    await putClasses(classes().filter((c) => c.name.toLowerCase() !== name.toLowerCase()));
  }

  async function setClassColor(name, color) {
    await putClasses(
      classes().map((c) => (c.name === name ? { ...c, color } : c))
    );
  }

  // Any label found in a loaded annotation that isn't in the project's class list gets
  // added, so the vocabulary reconstructs itself from the annotations. This is also the
  // safety net for the retired browser-local class list: nothing can be orphaned.
  async function ensureClasses(names) {
    if (!ACTIVE) return;
    const known = new Set(classNames().map((c) => c.toLowerCase()));
    const fresh = [];
    for (const n of names) {
      const t = String(n || "").trim();
      if (t && !known.has(t.toLowerCase())) {
        known.add(t.toLowerCase());
        fresh.push({ name: t, color: "" });
      }
    }
    if (fresh.length) await putClasses([...classes(), ...fresh]);
  }

  // --- Opening a project ------------------------------------------------------

  // The ONE code path used by boot, by switching, and by recovering from a delete.
  async function openProject(projectId) {
    let project = null;
    try {
      project = await API.getProject(projectId);
    } catch (err) {
      setProjStatus(err.message, true);
    }
    if (!project) {
      // Vanished (another tab deleted it, or a stale localStorage id). Re-enter boot.
      localStorage.removeItem(ACTIVE_KEY);
      await boot();
      setProjStatus(`project "${projectId}" no longer exists — opened another one`);
      return;
    }

    ACTIVE = project;
    API.setProject(project.id);
    localStorage.setItem(ACTIVE_KEY, project.id); // only AFTER a successful load
    renderPicker();
    renderPanel();

    // ---- PHASE 1: RESET. Every listener is a synchronous state reset. ----------
    document.dispatchEvent(
      new CustomEvent("slideprobe:project-opened", { detail: { project } })
    );
    // dispatchEvent has RETURNED, so every module has finished resetting.

    // ---- PHASE 2: WORK. Now, and only now, put a slide on screen. --------------
    window.SlideViewer.showProjectSlides(project, ALL_SLIDES);
  }

  function setStatus(msg) {
    const s = el("status");
    if (s) s.textContent = msg;
  }

  // --- Panel + picker ---------------------------------------------------------

  function renderPicker() {
    const picker = el("project-picker");
    if (!picker) return;
    picker.innerHTML = "";
    for (const p of PROJECTS) {
      const opt = document.createElement("option");
      opt.value = p.id;
      opt.textContent = p.name;
      picker.appendChild(opt);
    }
    if (ACTIVE) picker.value = ACTIVE.id;
  }

  function renderPanel() {
    renderProjectList();
    renderSlideChooser();
    const name = el("projects-active-name");
    if (name) name.textContent = ACTIVE ? ACTIVE.name : "—";
  }

  function renderProjectList() {
    const list = el("projects-list");
    if (!list) return;
    list.innerHTML = "";

    if (!PROJECTS.length) {
      const empty = document.createElement("div");
      empty.className = "prj-empty";
      empty.textContent = "No projects yet — create one below.";
      list.appendChild(empty);
      return;
    }

    for (const p of PROJECTS) {
      const row = document.createElement("div");
      row.className = "prj-row" + (ACTIVE && p.id === ACTIVE.id ? " active" : "");
      row.title = p.id; // two projects may share a name; the id disambiguates

      const name = document.createElement("span");
      name.className = "prj-name";
      name.textContent = p.name;
      name.addEventListener("click", () => {
        if (!ACTIVE || p.id !== ACTIVE.id) openProject(p.id);
      });

      const rename = document.createElement("button");
      rename.className = "prj-btn";
      rename.type = "button";
      rename.title = `rename "${p.name}"`;
      rename.textContent = "✎";
      rename.addEventListener("click", async () => {
        const next = prompt("Rename project", p.name);
        if (!next || next.trim() === p.name) return;
        try {
          await API.renameProject(p.id, next.trim());
          await refreshProjects();
          if (ACTIVE && ACTIVE.id === p.id) {
            ACTIVE.name = next.trim();
            renderPicker();
          }
          renderPanel();
        } catch (err) {
          setProjStatus(err.message, true);
        }
      });

      const del = document.createElement("button");
      del.className = "prj-btn prj-del";
      del.type = "button";
      del.title = `delete "${p.name}"`;
      del.textContent = "×";
      del.addEventListener("click", () => deleteProject(p));

      row.appendChild(name);
      row.appendChild(rename);
      row.appendChild(del);
      list.appendChild(row);
    }
  }

  async function deleteProject(p) {
    const ok = confirm(
      `Delete project "${p.name}"?\n\n` +
      "This removes its annotations, its classes and its trained model.\n" +
      "Your slides and their extracted features are NOT touched — any other " +
      "project over the same slides keeps working, and nothing has to be re-swept."
    );
    if (!ok) return;
    try {
      await API.deleteProject(p.id);
    } catch (err) {
      setProjStatus(err.message, true);
      return;
    }
    if (ACTIVE && ACTIVE.id === p.id) {
      // Drop the dead id BEFORE anything else can try to fetch with it.
      API.setProject(null);
      ACTIVE = null;
      localStorage.removeItem(ACTIVE_KEY);
      await boot();          // the same path as first load: pick another, or empty state
    } else {
      await refreshProjects();
      renderPanel();
    }
  }

  // Every slide on disk, checked if it's in this project. A slide that belongs to NO
  // project is still listed here — this chooser is the only way it can be adopted back
  // into one, which is what stops it from silently vanishing from the app.
  function renderSlideChooser() {
    const box = el("projects-slides");
    if (!box) return;
    box.innerHTML = "";
    if (!ACTIVE) return;

    const inProject = new Set(ACTIVE.slides);
    for (const s of ALL_SLIDES) {
      const row = document.createElement("label");
      row.className = "sl-row";

      const cb = document.createElement("input");
      cb.type = "checkbox";
      cb.checked = inProject.has(s.id);
      cb.addEventListener("change", () => toggleSlide(s.id, cb));

      const name = document.createElement("span");
      name.className = "sl-name";
      name.textContent = s.name;

      row.appendChild(cb);
      row.appendChild(name);
      box.appendChild(row);
    }

    // A slide listed in the project but no longer on disk: report it rather than
    // letting the viewer try to open a file that isn't there.
    const onDisk = new Set(ALL_SLIDES.map((s) => s.id));
    const missing = ACTIVE.slides.filter((id) => !onDisk.has(id));
    const outside = ALL_SLIDES.filter((s) => !inProject.has(s.id)).length;

    const notes = [];
    if (missing.length) notes.push(`${missing.length} missing from data/slides/`);
    if (outside) notes.push(`${outside} slide(s) on disk not in this project`);
    setProjStatus(notes.join(" · "));
  }

  async function toggleSlide(slideId, checkbox) {
    if (!ACTIVE) return;
    try {
      if (checkbox.checked) {
        await API.addProjectSlide(ACTIVE.id, slideId);
      } else {
        const res = await API.removeProjectSlide(ACTIVE.id, slideId, false);
        if (res.status === "has_annotations") {
          const ok = confirm(
            `"${slideId}" has annotations in this project.\n\n` +
            "Removing it will DELETE those annotations (only in this project — the " +
            "slide file and its extracted features stay).\n\nRemove it anyway?"
          );
          if (!ok) {
            checkbox.checked = true; // put the tick back; nothing happened
            return;
          }
          await API.removeProjectSlide(ACTIVE.id, slideId, true);
        }
      }
    } catch (err) {
      setProjStatus(err.message, true);
      checkbox.checked = !checkbox.checked; // the UI must not lie about the server
      return;
    }
    // Re-open the project so the slide picker and every module see the new membership.
    await refreshProjects();
    await openProject(ACTIVE.id);
  }

  async function refreshProjects() {
    PROJECTS = await API.listProjects();
    renderPicker();
  }

  // --- Empty state --------------------------------------------------------------

  function showEmptyState() {
    ACTIVE = null;
    API.setProject(null);
    localStorage.removeItem(ACTIVE_KEY);
    renderPicker();
    renderPanel();
    setStatus("no project — create one to start");
    window.SlidePanels.open("projects-panel");
    const input = el("projects-input");
    if (input) input.focus();
    // Nothing on screen: tell every module, so none of them keeps showing the last
    // project's slide state (features.js in particular would otherwise happily go on
    // reporting "✓ 4,281 tissue tiles" for a slide that is no longer open).
    window.SlideViewer.showNothing();
  }

  // --- Boot ---------------------------------------------------------------------

  async function boot() {
    try {
      [ALL_SLIDES, PROJECTS] = await Promise.all([
        API.listSlides(),
        API.listProjects(),
      ]);
    } catch (err) {
      setStatus(`could not reach backend: ${err.message}`);
      return;
    }

    if (!PROJECTS.length) {
      showEmptyState();
      return;
    }

    const stored = localStorage.getItem(ACTIVE_KEY);
    const known = PROJECTS.some((p) => p.id === stored);
    await openProject(known ? stored : PROJECTS[0].id);

    // Report the fallback AFTER opening, and in the Projects panel rather than in
    // #status: #status belongs to the viewer, which overwrites it with "showing <slide>"
    // as soon as the slide loads — so a notice written there is gone before it's read.
    if (stored && !known) {
      setProjStatus(`project "${stored}" no longer exists — opened "${PROJECTS[0].name}"`);
    }
  }

  // --- Wiring --------------------------------------------------------------------

  window.addEventListener("DOMContentLoaded", () => {
    // Phase 7: class labels moved server-side into project.json, where they carry an
    // explicit stored colour. Nothing on disk ever used the old browser-local list, so
    // there is nothing to migrate — drop the stale key rather than leave a ghost class
    // list in devtools. (And if any label DID exist, ensureClasses() above rebuilds the
    // vocabulary from the annotations themselves, so nothing can be lost.)
    localStorage.removeItem("slideprobe.customClasses");

    const btn = el("projects-btn");
    if (btn) btn.addEventListener("click", () => window.SlidePanels.toggle("projects-panel"));

    const close = el("projects-close");
    if (close) close.addEventListener("click", () => window.SlidePanels.close("projects-panel"));

    const picker = el("project-picker");
    if (picker) {
      picker.addEventListener("change", () => {
        if (picker.value && (!ACTIVE || picker.value !== ACTIVE.id)) openProject(picker.value);
      });
    }

    const form = el("projects-add");
    const input = el("projects-input");
    if (form && input) {
      form.addEventListener("submit", async (e) => {
        e.preventDefault();
        const name = input.value.trim();
        if (!name) return;
        try {
          const created = await API.createProject(name); // defaults to every slide on disk
          input.value = "";
          await refreshProjects();
          await openProject(created.id);
        } catch (err) {
          setProjStatus(err.message, true);
        }
      });
    }

    boot();
  });

  // What the rest of the frontend is allowed to know about the active project.
  window.SlideProject = {
    current: () => ACTIVE,
    id: () => (ACTIVE ? ACTIVE.id : null),
    classes,
    classNames,
    colorOf,
    addClass,
    removeClass,
    setClassColor,
    ensureClasses,
  };
})();
